### DISCUSSION TITLE
Add `allow-head-failure-until` input

### DISCUSSION THREAD

**Discussion by MSP-Greg** [ruby/setup-ruby#564] [Ideas]
Given the issues that occur with additions to bundled gems in ruby head builds, and (I hope) that many CI workflows are using Ruby head builds...

Add an input like `allow-head-failure-until: <some easily parsed date format>`, which setup-ruby could use to set an environment variable to be used in logic to allow head builds to fail, but only before the date?

IOW, an attempt to find a way to disable failures due to Ruby head builds, but only for a specified date period?  Or, maintainers disable them to stop CI failures, but it's easy to forget to reenable them...

This is a bit messy regarding TruffleRuby & JRuby, which probably don't have the issues with bundled gems (I think).  Maybe for now, don't include them in the 'date' calc?

**Comment by eregon** [comment]
Do such failures during the setup-ruby action?
Do you have an example log of such an issue?

If not during the setup-ruby action, of course we cannot influence other steps so that wouldn't work.

My other concern with this is there is no way to know when something will be fixed but the date would be mostly a guess and it's not unlikely it would fail again after that date.

**Comment by MSP-Greg** [comment]
> Do such failures during the setup-ruby action?

No.

The failures I'm speaking of are failures like the one you looked into in Puma (thank you).  Or, a std-lib item changing from default to bundled, causing build failures.  There may also be times when ruby head causes unforeseen issues.  Errors like
```
warning: base64 was loaded from the standard library, but is not part of the default gems since Ruby 3.4.0.
    Add base64 to your Gemfile or gemspec.
warning: bigdecimal was loaded from the standard library, but is not part of the default gems since Ruby 3.4.0.
    Add bigdecimal to your Gemfile or gemspec.
warning: mutex_m was loaded from the standard library, but is not part of the default gems since Ruby 3.4.0.
    Add mutex_m to your Gemfile or gemspec.
```

Two assumptions:
1. We want people to test with Ruby head, and we prefer they don't set CI to 'allow-failure'.
2. Assume that failures are occurring.  Currently, removing the job is the main option for getting CI passing.  It's easy to forget to re-enable the job at a later time.

I thought of this when I noticed Minitest's [`#skip_until`](https://msp-greg.github.io/ruby_trunk/minitest/Minitest/Assertions.html#skip_until-instance_method) method.

Using that syntax in Actions, re the Puma error, we could set `allow-head-failure-until: [2024, 2, 1]`, and the failing jobs would be ignored until 01-Feb-2024.  Hopefully that would be enough time for the dependencies to be sorted out...

**Reply by eregon** [reply]
> The failures I'm speaking of are failures like the one you looked into in Puma

Right so that's https://github.com/puma/puma/actions/runs/7529388305/job/20493515442 and discussed in https://github.com/puma/puma/pull/3320.

Even if setup-ruby had a `allow-head-failure-until` input, I don't see how it could prevent such a CI job to fail, because the step which fails is after setup-ruby is done, so setup-ruby cannot have any effect on that.
Or did I miss something?

Regarding the first assumption, I think using 'allow-failure' or more specifically `continue-on-error: true` *at the job level* on GitHub Actions is appropriate for head builds. It seems better than removing the job temporarily (and risk forgetting to add it back).
The UI is not ideal for sure, maybe one day GitHub will improve it.
For job-level `continue-on-error:`, the job is still shown red so at least it keeps awareness. Or IOW it doesn't have much effect from a PR POV.
For step-level `continue-on-error:` nothing is shown and one needs to expand the log to see if tests passed or not, that's bad.

An alternative, and this is what is done in concurrent-ruby is to move head builds to daily instead of PR CI: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/.github/workflows/experimental.yml

Yet another way is to keep the head CI job with the rest, but mark the rest as Required for PRs and the head one not as required, that way it should be clear that if the head job fail it might be OK and not necessarily a blocker to merge the PR.
I think that might be the best way currently.

**Reply by eregon** [reply]
> Using that syntax in Actions, re the Puma error, we could set `allow-head-failure-until: [2024, 2, 1]`, and the failing jobs would be ignored until 01-Feb-2024. Hopefully that would be enough time for the dependencies to be sorted out...

Ah maybe the idea is if current date < that value, then setup-ruby would use the latest Ruby release instead of ruby-head in that CI?
Seems a bit confusing and would duplicate another CI job, but might be a good idea nevertheless.

**Reply by MSP-Greg** [reply]
No, sorry, hopping between a few things.

All this would do is set an ENV variable like 'ALLOW_HEAD_FAILURE' so that later steps could determine whether to allow the CI job to fail, based on the date.

**Reply by eregon** [reply]
Ah I see, it'd be nice if https://docs.github.com/en/actions/learn-github-actions/expressions had a `now()` or something to do this directly in a step's `if:` or so.
Using allow-head-failure-until + using ALLOW_HEAD_FAILURE correctly seems not so easy.

**Reply by MSP-Greg** [reply]
Even with a now() or something, it wouldn't be aware of head builds.

For instance, with 3.2.3 released today, if it causes CI failures, I suspect most gems/repos would fix the problem quickly.  Not sure if the response would be the same for head builds.

> Using allow-head-failure-until + using ALLOW_HEAD_FAILURE correctly seems not so easy.

Let me look into it.  Can't recall using any js date functions recently, but I'll see if I can get it to work.  I can use the Puma workflow to test it...

**Reply by dentarg** [reply]
Sorry if off-topic, but re:

>For step-level `continue-on-error:` nothing is shown and one needs to expand the log to see if tests passed or not, that's bad.

failed jobs are listed in the "Summary" view, that combined with [logging the outcome of steps](https://github.com/puma/puma/blob/f6ab4e9b5fd3809acaab9407a79da4add7e77e31/.github/workflows/tests.yml#L203-L206) mitigates hunting for the error. I like this approach because the commits view will then be ✅ as that looks at individual jobs.

**Reply by eregon** [reply]
@dentarg Yes that's a useful trick. I think it still has the main downside though that if head builds are passing and a PR breaks them probably no one would notice (or much later) because one would need to open that job as it's not reflected at all on the job itself.

**Comment by dentarg** [comment]
I'm not sure this feature makes much sense / is worth it. I think all bets are off with ruby-head builds, here's an example where ruby-head was passing and then started to not pass: https://github.com/sinatra/sinatra/issues/1988

I get that we want exposure on ruby-head failures but to me it is not worth having the front page of your project show red failing builds. IMHO maintainers and interested contributors can check the Summary page from time to time to find out if head are causing problems (https://github.com/ruby/setup-ruby/discussions/564#discussioncomment-8178734).

**Reply by eregon** [reply]
> I'm not sure this feature makes much sense / is worth it. I think all bets are off with ruby-head builds, here's an example where ruby-head was passing and then started to not pass: [sinatra/sinatra#1988](https://github.com/sinatra/sinatra/issues/1988)

Yeah, I kinda agree, failure is somewhat expected with head builds, at least some part of the time for ruby-head, given incompatible changes for the next version.

> I get that we want exposure on ruby-head failures but to me it is not worth having the front page of your project show red failing builds.

Mmh, I see, and I guess moving head builds to daily doesn't actually solve that.
IMO it's fine to have head builds shown as red on the latest master commit and it's more annoying to show it as red on PRs as people might believe they broke something when it's not so likely.
